### PR TITLE
slate-react: MVP for working with non-global window objects

### PR DESCRIPTION
**Pull request by lukasbuenger**
#### Is this adding or improving a _feature_ or fixing a _bug_?

It's an improvement (I think)

#### What's the new behavior?

Since the rewrite, rendering `<Editable>` components into an `<iframe>` was no longer possible due to the `slate-react` package relying on the global `window` object for literally anything DOM-related: Event-handling, `instanceof` checks etc. 
With this PR, you can render your state into another `window.document` and everything should still work as expected. 

To retrieve the `window` object, on which a given editor instance is working, you'd call `ReactEditor.getWindow(editor)`. 

For a working example, check [this diff](#diff-e009c066107f606f805039682a84e53d) from below. It's also included in the [preview deploy of the website](https://deploy-preview-3819--slatejs.netlify.app), just select the example "Rendering in iframes" from the menu. To run it locally, `yarn build && yarn start` this PR and visit `http://localhost:3000/examples/iframes`. 

My main focus here was to put together a working MVP, which is a) fully backwards compatible and b) doesn't add to the existing public API surface (too much anyway).

##### Finding the "right" window object

Whenever the `ref` of an `<Editable>` DOM element becomes available, the related `window` is deduced by querying `ownerDocument.defaultView` and saved in a new WeakMap `EDITOR_TO_WINDOW`, with the editor instance as key. This allowed me to substitute any hard `window` calls from within `ReactEditor` methods (or wherever we have seamless access to the editor facade) with minimal hassle and stay somewhat inline with existing patterns. 

So instead of 
```
const domSelection = window.getSelection()
```
you would 
```
const window = ReactEditor.getWindow(editor)
const domSelection = window.getSelection()
```

Personally, I quite like this approach and the resulting API and it looks slate-ish enough to me.

##### Fixing `instanceof` checks

The most notable changes this PR introduces happen in the `utils/dom` module of `slate-react`. Every `window` instance comes with its own set of prototypes. This might not be super-obvious (especially in TS) but what `domNode instanceof Node` actually does is `domNode instanceof window.Node`. So, to safely assume the right prototype chain, I had to adjust some of the equality checks to work on arbitrary `window` objects. In order to not having to change/extend all the signatures of these methods, and given the fact that none of those helpers take an editor instance argument, I decided to go with something like this:

```
// isDOMNode?
!!(domNode &&
domNode.ownerDocument &&
domNode.ownerDocument.defaultView) &&
domNode instanceof domNode.ownerDocument.defaultView.Node
``` 

I'm not quite sure whether this approach is inline with the rest of the codebase and I also don't know to what extent such lookups impact overall performance, so please bear with me. It allowed me to keep the diff sensibly small, but I most definitely would appreciate any feedback in that regard.  

#### Have you checked that...?

- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: #3582, #3742 (these are the ones I found anyway)
Reviewers: @

### CODE REVIEW

**Comment by naripok** [comment]
Hey dude, I think you have a typo on the example file path.
On `site/pages/examples/[example].tsx`, it should be:

```
- ['Rendering in iframes', IFrames, 'iframes'],
+ ['Rendering in iframes', IFrames, 'iframe'],
``` 

Also, in order to see your example page, I had to modify `site/pages/api/index.ts`:
```
- export function getAllExamples() {
-   const slugs = readdirSync(examplePath)
-   return slugs
-     .filter(name => name.match(/.tsx$/))
-     .map(n => n.replace(/.tsx$/, ''))
- }
+ export function getAllExamples() {
+   const slugs = readdirSync(examplePath)
+   return slugs
+     .filter(name => name.match(/.(tsx|js)$/))
+     .map(n => n.replace(/.(tsx|js)$/, ''))
+ }
```
I could have changed the file extension to `tsx`, but my typechecker complains.

I have no previous experience with opensource and I don't want to mess it up, so I did not change it myself. Sorry.
Thanks for the fix, tho. It saved me a lot o work.

cheers!

File: site/pages/examples/[example].tsx
Comment (naripok) [inline site/pages/examples/[example].tsx:52]: Should be:
```
['Rendering in iframes', IFrames, 'iframe'],
```

**Review by naripok — CHANGES_REQUESTED** [review]
There is a typo on the file name on route discovery.

Also, https://github.com/ianstormtaylor/slate/blob/b616e75d636733733503c724c9084ff53e88dde8/site/pages/api/index.ts#L6-L11

should be changed to accept js files:
```
export function getAllExamples() {
  const slugs = readdirSync(examplePath)
  return slugs
    .filter(name => name.match(/.(tsx|js)$/))
    .map(n => n.replace(/.(tsx|js)$/, ''))
}
```
But I don't know if it is in the interest of the library maintainers.

Thx for the PR, tho. You saved me a lot of work!
Cheers!

**Comment by entiendoNull** [comment]
Any news on this? Working on upgrading Slate and migrate to the new version.
As I use Slate within an iframe a fix like this would surely help.

**Comment by naripok** [comment]
> Any news on this? Working on upgrading Slate and migrate to the new version.
> As I use Slate within an iframe a fix like this would surely help.

Hey dude,
I merged Lukas's PR on a fork and have been using it on a scoped package until the PR gets merged here.
It is working flawlessly for me until now.
It is hosted on github packages under `@naripok/slate`, if you want to check it out.

**Comment by entiendoNull** [comment]
Hey @naripok, thanks for the heads up.. unfortunately I did it myself before I saw your reply.
Works like a charm!

**Comment by lukasbuenger** [comment]
@naripok
Hi Fernando,
sorry for the late reply. I'm glad the PR works for you. Would you mind re-checking the example listing error, that you pointed out?
On a lower note, I'm really having a hard time understanding how this feature/behaviour was dropped during the rewrite in the first place, especially given the fact that it was relatively easy to implement. The infrastructure was already in place. 

Cheers, Lukas

**Review by naripok — APPROVED** [review]

**Comment by spudly** [comment]
Any chance this will be merged soon?

**Comment by spudly** [comment]
pretty please? 🍒

**Comment by Morphexe** [comment]
Would love to get this merged in aswell :)

**Review by ianstormtaylor — APPROVED** [review]

**Comment by Mangatt** [comment]
@thesunny Any chance to merge that since it's already aproved by @ianstormtaylor ?

**Comment by thesunny** [comment]
@Mangatt I've set aside some time on Wednesday between meetings to review and merge. Wednesday is my "meetings and other stuff" day. If all is good, I should have it merged then. The actual merge takes hardly any time, but I need to reserve time around it in case of something unexpected.

**Comment by Mangatt** [comment]
That would be great, thank you @thesunny !

**Comment by thesunny** [comment]
I didn't know how to contribute to the PR and keep the attributions pointing to the right place so I created a new PR:

#4079 

Will merge the other one.

**Comment by thesunny** [comment]
Published in 0.60.4 which is a "@next" version

**Comment by Mangatt** [comment]
It works great @thesunny , thanks!

**Comment by lukasbuenger** [comment]
@thesunny Thank you so much for the release, this is really appreciated. I had hacky variations of this running in (low-scale, Chrome only) prod ever since the rewrite. As of now, the @next release works flawlessly for me.

**Pull request closed without merging**
